src/magic_folder/tahoe_client.py

```diff
@@ -0,0 +1,229 @@
+
+        :param producer: can take anything that treq's data= method to
+            treq.request allows which is currently: str, file-like or
+            IBodyProducer. See
+            https://treq.readthedocs.io/en/release-20.3.0/api.html#treq.request
+
+        :return Deferred[bytes]: A Deferred which fires with the capability
+            string for the new immutable object.
+        """
+        put_uri = self.url.replace(
+            path=(u"uri",),
+            query=[(u"mutable", u"false")],
+        )
+        res = yield _request(
+            self.http_client,
+            b"PUT",
+            put_uri,
+            data=producer,
+        )
+        capability_string = yield _get_content_check_code({CREATED}, res)
+        returnValue(
+            capability_string.strip()
+        )
+
+    @inlineCallbacks
+    def create_mutable_directory(self):
+        """
+        Create a new mutable directory in Tahoe.
+
+        :return Deferred[bytes]: The write capability string for the new
+            directory.
+        """
+        post_uri = self.url.replace(
+            path=(u"uri",),
+            query=[(u"t", u"mkdir")],
+        )
+        response = yield _request(
+            self.http_client,
+            b"POST",
+            post_uri,
+        )
+        # Response code should probably be CREATED but it seems to be OK
+        # instead.  Not sure if this is the real Tahoe-LAFS behavior or an
+        # artifact of the test double.
+        capability_string = yield _get_content_check_code({OK, CREATED}, response)
+        returnValue(capability_string)
+
+    @inlineCallbacks
+    def download_capability(self, cap):
+        """
+        Retrieve the raw data for a capability from Tahoe
+
+        :param cap: a capability-string
+
+        :returns: bytes
+        """
+        get_uri = self.url.replace(
+            path=(u"uri",),
+            query=[(u"uri", cap.decode("ascii"))],
+        )
+        res = yield _request(
+            self.http_client,
+            b"GET",
+            get_uri,
+        )
+        data = yield _get_content_check_code({OK}, res)
+        returnValue(data)
+
+    @inlineCallbacks
+    def stream_capability(self, cap, filelike):
+        """
+        Retrieve the raw data for a capability from Tahoe
+
+        :param cap: a capability-string
+
+        :param filelike: a writable file object. `.write` will be
+            called on it an arbitrary number of times, but no other
+            methods (that is, it won't be closed).
+
+        :returns: Deferred that fires with `None`
+        """
+        get_uri = self.url.replace(
+            path=(u"uri",),
+            query=[(u"uri", cap.decode("ascii"))],
+        )
+        res = yield self.http_client.get(get_uri.to_text())
+        if res.code != OK:
+            raise TahoeAPIError(res.code, None)
+        yield res.collect(filelike.write)
+
+
+def create_tahoe_client(url, http_client):
```

Comment:
When this expands to need the token then accepting a `node_directory` might make sense.  I suspect we won't really care about that for a long time though.  I'm not sure if it should ever take a `GlobalConfigDatabase` though ... maybe tahoe_client also belongs in Tahoe-LAFS itself, or another stand-alone package (since it isn't really tied to magic-folder).